File: compiler/nodes/statements/trait_definition.go
package statements

import (
	"wipple/codegen"
	"wipple/database"
	"wipple/nodes/attributes"
	"wipple/nodes/constraints"
	"wipple/nodes/types"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type TraitDefinitionNode struct {
	Comments    []string
	Attributes  []*attributes.AttributeNode
	Name        string
	Parameters  []database.Node
	Type        database.Node
	Constraints []database.Node
	Facts       *database.Facts
}

func (node *TraitDefinitionNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseTraitDefinitionStatement(parser *syntax.Parser) (*TraitDefinitionNode, *syntax.Error) {
	span := parser.Spanned()

	comments, err := ParseComments(parser)
	if err != nil {
		return nil, err
	}

	attributes, err := attributes.ParseAttributes(parser)
	if err != nil {
		return nil, err
	}

	name, err := syntax.ParseTypeName(parser)
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("AssignOperator")
	if err != nil {
		return nil, err
	}

	parser.ConsumeLineBreaks()

	parameters, err := types.ParseTypeParameters(parser)
	if err != nil {
		return nil, err
	}

	ty, constraints, err := ParseTraitConstraints(parser)
	if err != nil {
		return nil, err
	}

	return &TraitDefinitionNode{
		Comments:    comments,
		Attributes:  attributes,
		Name:        name,
		Parameters:  typeParameterNodes(parameters),
		Type:        ty,
		Constraints: constraints,
		Facts:       database.NewFacts(span()),
	}, nil
}

func ParseTraitConstraints(parser *syntax.Parser) (database.Node, []database.Node, *syntax.Error) {
	_, err := parser.Token("TraitKeyword")
	if err != nil {
		return nil, nil, err
	}

	parser.Commit("in this trait definition")

	traitType, err := types.ParseAtomicType(parser)
	if err != nil {
		return nil, nil, err
	}

	constraints, _, err := syntax.ParseOptional(parser, constraints.ParseConstraints)
	if err != nil {
		return nil, nil, err
	}

	return traitType, constraints, nil
}

func (node *TraitDefinitionNode) Visit(visitor *visit.Visitor) {
	visit.Defining(visitor, node, func() (*visit.TraitDefinition, bool) {
		visitor.PushScope()

		visitor.CurrentDefinition.WithImplicitTypeParameters(func() {
			for _, parameter := range node.Parameters {
				visitor.Visit(parameter)
			}
		})

		visitor.AfterAllDefinitions(func() {
			visitor.Visit(node.Type)

			visitor.Constraint(typecheck.NewGroupConstraint(node, node.Type))

			for _, constraint := range node.Constraints {
				visitor.Visit(constraint)
			}

			// The bound for this trait is added where needed
		})

		visitor.PopScope()

		definition := &visit.TraitDefinition{
			Name:       node.Name,
			Node:       node,
			Comments:   node.Comments,
			Attributes: visit.ParseTraitAttributes(node.Attributes),
			Parameters: node.Parameters,
		}

		visitor.Define(node.Name, definition)

		return definition, true
	})
}

func (node *TraitDefinitionNode) Codegen(c *codegen.Codegen) error {
	// Handled specially in `codegen.Codegen`
	return nil
}

File: compiler/src/nodes/statements/trait_definition.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenError},
    database::{Node, NodeRef},
    nodes::{
        parse_atomic_type, parse_attributes, parse_comments, parse_constraints,
        parse_type_parameters,
    },
    syntax::{ParseError, Parser, TokenKind, parse_type_name},
    typecheck::GroupConstraint,
    visit::{TraitAttributes, TraitDefinition, Visit, Visitor},
};

#[derive(Debug)]
pub struct TraitDefinitionNode {
    pub comments: Vec<String>,
    pub attributes: Vec<NodeRef>,
    pub name: String,
    pub parameters: Vec<NodeRef>,
    pub ty: NodeRef,
    pub constraints: Vec<NodeRef>,
}

impl Node for TraitDefinitionNode {}

pub fn parse_trait_definition_statement(
    parser: &mut Parser<'_>,
) -> Result<TraitDefinitionNode, ParseError> {
    let comments = parse_comments(parser)?;
    let attributes = parse_attributes(parser)?;
    let name = parse_type_name(parser)?;

    parser.token(TokenKind::AssignOperator)?;
    parser.consume_line_breaks();

    let parameters = parse_type_parameters(parser)?;
    let (ty, constraints) = parse_trait_constraints(parser)?;

    Ok(TraitDefinitionNode {
        comments,
        attributes,
        name,
        parameters,
        ty,
        constraints,
    })
}

pub fn parse_trait_constraints(
    parser: &mut Parser<'_>,
) -> Result<(NodeRef, Vec<NodeRef>), ParseError> {
    parser.token(TokenKind::TraitKeyword)?;
    parser.commit("in this trait definition");

    let ty = parse_atomic_type(parser)?;
    let constraints = parser
        .parse_optional(parse_constraints)?
        .unwrap_or_default();

    Ok((ty, constraints))
}

impl Visit for TraitDefinitionNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visitor.defining(node, |visitor| {
            visitor.push_scope();

            visitor.with_implicit_type_parameters(|visitor| {
                for parameter in &self.parameters {
                    visitor.visit(parameter);
                    visitor.edge(parameter, node, "parameter");
                }
            });

            let ty = self.ty.clone();
            let constraints = self.constraints.clone();
            let node_for_closure = node.clone();

            visitor.after_all_definitions(move |visitor| {
                let node = node_for_closure.clone();
                visitor.visit(&ty);
                visitor.edge(&ty, &node, "type");

                visitor.constraint(GroupConstraint::new(node.clone(), ty.clone()));

                for constraint in &constraints {
                    visitor.visit(constraint);
                    visitor.edge(constraint, &node, "constraint");
                }
            });

            visitor.pop_scope();

            let definition = TraitDefinition {
                name: self.name.clone(),
                node: node.clone(),
                comments: self.comments.clone(),
                attributes: TraitAttributes {},
                parameters: self.parameters.clone(),
            };

            visitor.define(&definition.name, definition.clone());

            definition
        });
    }
}

impl Codegen for TraitDefinitionNode {
    fn codegen(&self, _codegen: &mut CodegenCtx<'_>) -> Result<(), CodegenError> {
        Ok(())
    }
}
